### ISSUE TITLE
Get Representation instead of StateObjectSelector

### ISSUE THREAD

**Issue by htaldo** [molstar/molstar#1074]
Good morning!
I am trying a combination of #168 and #989 to save references to representations in an array. I added the update and setState lines for testing purposes:
```typescript
async function createPocketRepr (pocket: Pocket) {
	loadedReprs[pocket.rank] = [];
	for (const item of pocket.seq) {
	    let chain = item.chain;
	    let subSeq = item.subSeq; //an array holding auth_seq_ids
	    let query = createSequenceIdExpression(chain, subSeq);
	    let component = await plugin.builders.structure.tryCreateComponentFromExpression(currentStructure, query, 'component');
	    const repr = await plugin.builders.structure.representation.addRepresentation(component!, {
		type: 'molecular-surface',
		typeParams: { alpha: 0.5 },
		color: 'uniform',
		colorParams: { value: Color(0x86F1E9) }
	    });
	    plugin.canvas3d.update(repr);
	    loadedReprs[pocket.rank].push(repr);
	    repr.setState({visible: false});
	}
    }
```
    
I think the representations are actually being generated, since they do appear in the viewer with the requested options:
![image](https://github.com/molstar/molstar/assets/64457490/1ffb6221-8527-4061-beff-8cd13ea50b61)
    
However, my LSP/npm throws these errors:
On canvas.update:
    `TS2345: Argument of type 'StateObjectSelector<Representation3D, Transformer<StateObject<any, Type<any>>, StateObject<any, Type<any>>, any>>' is not assignable to parameter of type 'Any<Params, State>'.
  Type 'StateObjectSelector<Representation3D, Transformer<StateObject<any, Type<any>>, StateObject<any, Type<any>>, any>>' is missing the following properties from type 'Representation<any, Params, State>': label, updated, groupCount, renderObjects, and 12 more.`
On repr.setState:
`Property 'setState' does not exist on type 'StateObjectSelector<Representation3D, Transformer<StateObject<any, Type<any>>, StateObject<any, Type<any>>, any>>'.`

Trying a conversion of repr as Repr.Any results in a similar error, so I guess that's a type mismatch.
I am using repr.setState because I was trying to toggle the visibility from an external function, so that's where I first encountered the error.
How could I get the reference to the generated representations, instead of StateObjectSelectors?  

P.D.: Not sure if it's related, but adding more representations results in the previous one disappearing, so only one pocket is displayed at once. Unfortunately, I lost access to the representations in the UI, so I can't tell how they are changing from there. Sorry for the lengthy post and thanks in advance!

**Comment by htaldo** [comment]
After some digging up, I found in #404 how to update the representation using the selector (my mistake was that I thought the API worked like stefdoerr mentioned in #557). Now I can turn off the representation by making it invisible :
```typescript
const pocketReprOff = (pocket: Pocket) => {
	for (const repr of loadedReprs[pocket.rank]) { //repr is a StateObjectSelector
	    plugin.build()
		.to(repr)
		.update(createStructureRepresentationParams(plugin, undefined, {
		    type: 'molecular-surface',
		    typeParams: { alpha: 0 },
		    color: 'uniform',
		    colorParams: { value: Color(0x86F1E9) }
	    }))
	    .commit();
	}
    }
```

But the representations are still being replaced by the latest one with successive calls of createPocketRepr. By doing some logging I found the component ref doesn't change through the iterations of the loop, so it looks like each new representation "overwrites" the former one. I also tried making an array of components to try to get a unique reference in each iteration, but the result was the same.
```typescript
async function createPocketRepr (pocket: Pocket) {
	loadedReprs[pocket.rank] = [];
	for (const item of pocket.seq) {
	    let chain = item.chain;
	    let subSeq = item.subSeq;
	    let query = createSequenceIdExpression(chain, subSeq);
	    let component = await plugin.builders.structure.tryCreateComponentFromExpression(currentStructure, query, 'pocket${pocket.rank}');
	    const selector = await plugin.builders.structure.representation.addRepresentation(component, {
		type: 'molecular-surface',
		typeParams: { alpha: 0.5 },
		color: 'uniform',
		colorParams: { value: Color(0x86F1E9) }
	    });
	    loadedReprs[pocket.rank].push(selector);
	}
    }
```
I'll keep trying to find what I'm missing there

**Comment by dsehnal** [comment]
Hi,

some pointers:

```ts
const pocketReprOff = (pocket: Pocket) => {
	for (const repr of loadedReprs[pocket.rank]) { //repr is a StateObjectSelector
	    plugin.build()
		.to(repr)
		.update(createStructureRepresentationParams(plugin, undefined, {
		    type: 'molecular-surface',
		    typeParams: { alpha: 0 },
		    color: 'uniform',
		    colorParams: { value: Color(0x86F1E9) }
	    }))
	    .commit();
	}
    }
```

This is wrong as the commit returns a promise and the update is asychronous. So you are essentially running all the updates in parallel and they end up overwriting each other. If you are doing multiple updates at the same time, you want something like this:

```ts
const updates = plugin.build();
for (const x in xs) {
  updates.to(x).update(...);
}
await updates.commit();
```

-------------

You are already storing the refs/selectors. To hide a visual, you can use ``setSubtreeVisibility(plugin.state.data, cell.ref, false)`` instead of manipulating the opacity.

-------------

In your original post, you are calling `plugin.canvas3d.update(repr);`, but that is called for you automatically by the plugin runtime as it syncs the state tree with the canvas.

**Comment by htaldo** [comment]
@dsehnal thanks for taking the time and the comprehensive response!
I moved the await line out of the loop, but I'm still pretty much facing the same problem, though I might have fixed one that would show up later. With this hint, I do think it has something to do with how promises are being handled in createPocketRepr, so I'll take a deeper look into that!

**Comment by htaldo** [comment]
Turns out, the strange behavior in createPocketRepr was because I was using the same key arg for tryCreateComponentFromExpression (at some point I changed it to be my key variable, but I forgot to use backticks and used single quotes instead, so it was picking the same name all along).
Now it's all working smoothly. I'll consider merging the components of each chain into a single pocket component, though.
Thanks again @dsehnal! I'm eager to keep learning more from this library!

**Comment by htaldo** [comment]
Turns out, the strange behavior in createPocketRepr was because I was using the same key arg for tryCreateComponentFromExpression (at some point I changed it to be my key variable, but I forgot to use backticks and used single quotes instead, so it was picking the same name all along).
Now it's all working smoothly. I'll consider merging the components of each chain into a single pocket component, though.
Thanks again @dsehnal! I'm eager to keep learning more from this library!
